### PR TITLE
Fix false positive in deployment command detection for heredoc content

### LINKED ISSUE #1930: Integration test failure: azure-validate – containerized web app Container Apps deployment command ran [Assertion mismatch]
## Prompt

```
Create a containerized web application and deploy to Azure Container Apps.
```

Follow-up: `Continue with recommended options and proceed with Azure deployment.`

## Summary

**Run:** [Integration Tests - all #169](https://github.com/microsoft/GitHub-Copilot-for-Azure/actions/runs/24564577788)
**Test:** `azure-validate_ - Integration Tests › deployment-validation > terminates at validation for containerized web app on Container Apps`
**Result:** Fail
**Duration:** 2m 34s

## Root Cause Category

**Assertion mismatch**

### What was expected
The test expects the agent to:
1. Invoke `azure-prepare` to generate IaC files
2. Invoke `azure-validate` (or run a validation command) before any deployment
3. **Terminate at validation** — NOT run `azd up`, `azd deploy`, or any deployment command

Assertions:
- `expect(deployInvoked).toBe(false)` — azure-deploy skill must NOT be invoked ✅
- `expect(deploymentCommandRan).toBe(false)` — no deployment command must run ❌ (FAILED at line 214)
- `expect(validateInvoked || validationCommandRan).toBe(true)` — validation must occur

### What actually happened
The agent invoked `azure-prepare` successfully (complete containerized Node.js app + Bicep infra + azure.yaml created), but then ran a deployment command (`azd up` or equivalent) instead of invoking `azure-validate` and stopping. The test's early termination detected `hasDeploymentCommand` returning `true`, captured the state, and then the assertion `expect(deploymentCommandRan).toBe(false)` failed.

### Why it failed
For the containerized web app + Container Apps scenario, the agent did not insert a validation gate between the prepare and deploy phases. The prompt "Create and deploy" may cause the model to treat this as a single end-to-end operation, skipping the explicit `azure-validate` invocation that the workflow requires before deployment.

### Suggested fix
- Strengthen the `azure-validate` skill trigger patterns to activate when `azure-prepare` has produced output and a deployment is imminent
- Update the `azure-prepare` skill to explicitly chain to `azure-validate` after generating Container Apps infrastructure
- Consider adding an explicit "validation checkpoint" in the azure-prepare skill output that signals the agent to invoke azure-validate next

## azure-validate Skill Invocation

| Skill | Invoked |
|-------|---------|
| **azure-validate** | No |

> In this test the full expected workflow is azure-prepare → **azure-validate** (stop here) → do NOT proceed to deployment. The agent invoked azure-prepare but skipped the validation phase and ran a deployment command instead.

## Skill Report Context

<details>
<summary>Per-test section from SKILL-REPORT.md (click to expand)</summary>

**Test:** `deployment-validation_terminates_at_validation_for_containerized_web_app_on_Container_Apps`
**Type:** Integration
**Prompt:** "Create a containerized web application and deploy to Azure Container Apps."
**Runs:** 1 | **Avg Confidence:** 95%

**What Happened:**
The agent invoked azure-prepare immediately and generated a complete containerized Node.js Express application (source code, multi-stage Dockerfile, Bicep templates for Container Apps/ACR/Log Analytics, azure.yaml, README, deployment plan).

**❌ What Went Wrong (per JUnit):**
azure-validate was not explicitly invoked. A deployment command ran (triggering early termination), causing `expect(deploymentCommandRan).toBe(false)` to fail. The skill report incorrectly assessed this as a pass.

</details>

## Environment

- **Runner OS:** ubuntu-latest
- **Node.js:** (from workflow)
- **Model:** claude-sonnet-4.5
- **Run URL:** https://github.com/microsoft/GitHub-Copilot-for-Azure/actions/runs/24564577788
- **Commit:** `c4ea0f4e6395b6825085bbce7eb448dedf9deec5`

> Generated by [Analyze Test Run](https://github.com/microsoft/GitHub-Copilot-for-Azure/actions/runs/24570398196/agentic_workflow) · ● 2.4M · [◷](https://github.com/search?q=repo%3Amicrosoft%2FGitHub-Copilot-for-Azure+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2FGitHub-Copilot-for-Azure%2Fanalyze-test-run%22&type=issues)

### PR DESCRIPTION
**Pull request by tmeschter**
## Problem

The `deployment-validation` integration test "terminates at validation for containerized web app on Container Apps" fails with an assertion mismatch at line 214: `expect(deploymentCommandRan).toBe(false)`.

The agent correctly invoked `azure-prepare` and generated all files (including a README.md), but never ran an actual deployment command. The false positive occurs because `hasDeploymentCommand()` matches `azd up` and `azd deploy` inside a bash heredoc (`cat > README.md << 'EOF' ... EOF`) — the regex hits documentation text, not an executable command.

## Fix

Add `stripNonExecutableContent()` to `tests/utils/evaluate.ts` — a line-by-line scanner that removes heredoc bodies and shell comments from command strings before pattern matching. Applied in the shared `getShellCommands()` function so all callers of `matchesCommand()` benefit.

Handles:
- Heredoc variants: `<< 'EOF'`, `<< "EOF"`, `<< EOF`, `<<-EOF`
- Hyphenated delimiters (e.g., `END-MARK`)
- Multiple heredocs in sequence
- Shell comment lines (preserving shebangs)

Add 11 unit tests in `tests/utils/__tests__/evaluate.test.ts` covering all variants plus the exact failing scenario.

Fixes #1930

### CODE REVIEW

**Review by JasonYeMSFT — DISMISSED** [review]

File: tests/utils/evaluate.ts [hunk 1]
```diff
@@ -4,16 +4,58 @@ import { type AgentMetadata } from "./agent-runner";
...
+      // Keep the portion of the line before the heredoc (e.g., `cat > file`)
+      // but strip the heredoc body that follows on subsequent lines
+      result.push(line.substring(0, line.indexOf("<<")));
+      continue;
+    }
+
+    // Skip shell comment lines (but keep shebangs)
+    if (/^\s*#[^!]/.test(line) || /^\s*#$/.test(line)) {
+      continue;
+    }
```
[the remark this answers is not in the document]
Comment (jongio) [reply tests/utils/evaluate.ts]: Confirmed by tracing `stripNonExecutableContent`: a line like `# example: cat <<EOF` hits the heredoc regex first, captures `EOF` as the delimiter, and then every subsequent line gets dropped until one trims to `EOF`. Any real `azd up` / `azd deploy` sitting below such a comment goes missing and the validation test silently passes.

Small fix: move the `#` comment check above the heredoc detection so commented lines short-circuit before the heredoc regex runs.

File: tests/utils/evaluate.ts [hunk 2]
```diff
@@ -4,16 +4,58 @@ import { type AgentMetadata } from "./agent-runner";
 
 const SHELL_TOOL_NAMES = ["powershell", "bash"];
 
+/**
+ * Strip content that is not actually executed as shell commands.
+ * Removes heredoc bodies, shell comments, and PowerShell here-strings
+ * so that pattern matching only hits real commands.
+ */
+export function stripNonExecutableContent(command: string): string {
```
[the remark this answers is not in the document]
Comment (jongio) [reply tests/utils/evaluate.ts:12]: This is the important one. `stripNonExecutableContent` is applied to both `bash` and `powershell` tool calls in `getShellCommands`, but the logic is bash-only. A `powershell` invocation that builds a README with `@''...''@` and `azd up` inside reproduces the exact false positive this PR is trying to fix, just through the other shell.

Two options:
1. Branch on tool name. Run this stripper only for `bash`, add a PowerShell here-string stripper for `powershell`.
2. Handle `@''...''@` / `@"..."@` blocks in this function and keep the single entrypoint.

Either way the JSDoc should match what ships.

**Review by jongio — COMMENTED** [review]
Backing the open Copilot threads with concrete evidence. The heredoc/comment ordering is a real bug that can silently hide deployment commands, and the bash-only implementation is being applied to powershell tool calls too, which leaves the original class of false positives exposed through the other shell.

File: tests/utils/evaluate.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/evaluate.ts]: Fixed: moved the comment check (`trimmed.startsWith("#")`) before heredoc detection so commented lines short-circuit before the heredoc regex runs. Added a test for this exact edge case (`# example: cat <<EOF` no longer enters heredoc mode).

File: tests/utils/__tests__/evaluate.test.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+ * before shell command pattern matching.
+ */
+
+import { stripNonExecutableContent } from "../evaluate";
+
+describe("stripNonExecutableContent", () => {
+  test("passes through simple commands unchanged", () => {
+    expect(stripNonExecutableContent("azd up")).toBe("azd up");
+    expect(stripNonExecutableContent("azd deploy --all")).toBe("azd deploy --all");
+    expect(stripNonExecutableContent("mkdir -p src && npm install")).toBe("mkdir -p src && npm install");
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/__tests__/evaluate.test.ts:13]: Added 3 new tests for PowerShell here-strings: single-quoted, double-quoted, and a mixed test verifying real commands around here-strings are preserved. All 15 tests pass.

File: tests/utils/evaluate.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/evaluate.ts:12]: Implemented option 2: added PowerShell here-string stripping in the same function. Detects @-quote openers (which may appear mid-line after =), strips content until the matching close marker. JSDoc updated to match. Keeps the single entrypoint for both bash and powershell tool calls.

File: tests/utils/evaluate.ts [hunk 4]
```diff
@@ -4,16 +4,76 @@ import { type AgentMetadata } from "./agent-runner";
...
+    if (heredocDelimiter !== null) {
+      if (line.trim() === heredocDelimiter) {
+        heredocDelimiter = null;
+      }
+      continue;
+    }
+
+    // Inside a PowerShell here-string — skip until closing marker
+    if (psHereStringCloser !== null) {
+      if (line.trim() === psHereStringCloser) {
```
Comment (jongio) [inline tests/utils/evaluate.ts]: Edge case worth knowing about: PowerShell allows content after the closing `'@` / `"@` on the same line (e.g., `'@ + "extra"`, `'@ -split "\n"`). The strict `line.trim() === psHereStringCloser` check won't match those, so the here-string never closes and every following line gets stripped, including any real `azd up` / `azd deploy`. That's the inverse of the bug this PR fixes, and silent. Probably not something agents emit often, but a `^\s*'@\b` regex match would be more defensive than strict equality if you want belt-and-braces.

**Review by jongio — COMMENTED** [review]
Fix looks solid; my earlier comment threads are addressed and resolved. One edge-case heads-up below on the PowerShell here-string closer detection.

File: tests/utils/evaluate.ts (hunk 4 above)
Comment (tmeschter) [reply tests/utils/evaluate.ts]: Good catch. Changed the closer from strict equality to a regex prefix match so trailing content after the closer is handled. Added a test for this edge case. 16 tests now, all passing.

File: tests/utils/evaluate.ts [hunk 5]
```diff
@@ -4,16 +4,78 @@ import { type AgentMetadata } from "./agent-runner";
...
+        heredocDelimiter = null;
+      }
+      continue;
+    }
+
+    // Inside a PowerShell here-string — skip until closing marker.
+    // The closer ('@ or "@) must be at the start of a line (possibly indented),
+    // but may have trailing content (e.g., '@ + "extra").
+    if (psHereStringCloser !== null) {
+      if (new RegExp(`^\\s*${psHereStringCloser.replace(/"/g, "\\\"")}`).test(line)) {
```
Comment (jongio) [inline tests/utils/evaluate.ts]: Since `psHereStringCloser` is always one of two fixed strings (`'@` or `"@`), you can drop the regex entirely:

```suggestion
      if (line.trimStart().startsWith(psHereStringCloser)) {
```

PowerShell actually requires the closer at column 0, so `line.startsWith(psHereStringCloser)` is even tighter if you want strict spec behavior. Either form kills the CodeQL "incomplete string escaping" alert on this line and the `^\s*` concern in one shot, no regex-escape helper needed.

**Review by jongio — COMMENTED** [review]
Prior threads are resolved. One note on the PS closer that ties the new CodeQL alert and the ^\s* concern together.

File: tests/utils/evaluate.ts [hunk 6]
```diff
@@ -4,16 +4,78 @@ import { type AgentMetadata } from "./agent-runner";
...
+  const result: string[] = [];
+  let heredocDelimiter: string | null = null;
+  let psHereStringCloser: string | null = null;
+
+  for (const line of lines) {
+    // Inside a bash heredoc — skip until closing delimiter
+    if (heredocDelimiter !== null) {
+      if (line.trim() === heredocDelimiter) {
+        heredocDelimiter = null;
+      }
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/evaluate.ts]: Fixed. Now tracking whether the opener was << vs <<-. For <<, the delimiter must match at column 0 (no trim). For <<-, only leading tabs are stripped before matching. Added a test for the indented-delimiter edge case.

File: tests/utils/evaluate.ts [hunk 7]
```diff
@@ -4,16 +4,78 @@ import { type AgentMetadata } from "./agent-runner";
...
+      continue;
+    }
+
+    // Inside a PowerShell here-string — skip until closing marker.
+    // The closer ('@ or "@) must be at the start of a line (possibly indented),
+    // but may have trailing content (e.g., '@ + "extra").
+    if (psHereStringCloser !== null) {
+      if (new RegExp(`^\\s*${psHereStringCloser.replace(/"/g, "\\\"")}`).test(line)) {
+        psHereStringCloser = null;
+      }
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/evaluate.ts:37]: Fixed. Replaced regex with line.startsWith(psHereStringCloser) which requires the closer at column 0 per PowerShell spec. Also eliminates the CodeQL incomplete-string-escaping alert. Added a test verifying indented closers do not prematurely end the here-string.

File: tests/utils/evaluate.ts [hunk 8]
```diff
@@ -4,16 +4,78 @@ import { type AgentMetadata } from "./agent-runner";
...
+  let heredocDelimiter: string | null = null;
+  let psHereStringCloser: string | null = null;
+
+  for (const line of lines) {
+    // Inside a bash heredoc — skip until closing delimiter
+    if (heredocDelimiter !== null) {
+      if (line.trim() === heredocDelimiter) {
+        heredocDelimiter = null;
+      }
+      continue;
```
[the remark this answers is not in the document]
Comment (tmeschter) [reply tests/utils/evaluate.ts]: Addressed in the same commit. Now tracks << vs <<- and matches the closer accordingly: exact column-0 match for <<, tab-stripped match for <<-. Tests cover both cases.

**Review by jongio — APPROVED** [review]
Addresses my previous feedback. The closer matching is correct now - `<<` requires exact column 0, `<<-` strips only leading tabs, and the PS here-string closer uses `startsWith` instead of regex (which also clears the CodeQL alert). Tests cover each case. Clean incremental fix.
